File: Python/Tema9_Ejercicio8.py
"""
TEMA 9: PROGRAMACIÓN ORIENTADA A OBJETOS
EJERCICIO 8: CLASES ABSTRACTAS
================================

Define interfaces y contratos de clases
"""

# CLASES ABSTRACTAS BÁSICAS
print("=== CLASES ABSTRACTAS ===\n")

from abc import ABC, abstractmethod

class Pagador(ABC):
    @abstractmethod
    def pagar(self, cantidad):
        pass

class TarjetaCredito(Pagador):
    def __init__(self, numero):
        self.numero = numero
    
    def pagar(self, cantidad):
        return f"Pagando ${cantidad} con tarjeta {self.numero}"

class PayPal(Pagador):
    def __init__(self, email):
        self.email = email
    
    def pagar(self, cantidad):
        return f"Pagando ${cantidad} mediante PayPal ({self.email})"

class Bitcoin(Pagador):
    def __init__(self, billetera):
        self.billetera = billetera
    
    def pagar(self, cantidad):
        return f"Pagando ${cantidad} en Bitcoin a {self.billetera}"

# PROCESAR PAGOS
print("Procesando pagos:")

pagadores = [
    TarjetaCredito("1234-5678-9012-3456"),
    PayPal("usuario@gmail.com"),
    Bitcoin("1A1z7agoat2AxWLcHE8yyQkqesSqrE9Vc")
]

for pagador in pagadores:
    print(f"  {pagador.pagar(100)}")

# MÉTODO ABSTRACTO CON IMPLEMENTACIÓN
print("\n=== CON IMPLEMENTACIÓN ===\n")

class Vehículo(ABC):
    def __init__(self, marca):
        self.marca = marca
    
    @abstractmethod
    def acelerar(self):
        pass
    
    def obtener_marca(self):
        return self.marca

class Coche(Vehículo):
    def acelerar(self):
        return f"{self.marca}: ¡Brrrr!"

class Bicicleta(Vehículo):
    def acelerar(self):
        return f"{self.marca}: Pedalea fuerte"

vehiculos = [Coche("Ferrari"), Bicicleta("Trek")]
for v in vehiculos:
    print(f"{v.obtener_marca()}: {v.acelerar()}")

"""
SALIDA ESPERADA:
=== CLASES ABSTRACTAS ===

Procesando pagos:
  Pagando $100 con tarjeta 1234-5678-9012-3456
  Pagando $100 mediante PayPal (usuario@gmail.com)
  Pagando $100 en Bitcoin a 1A1z7agoat2AxWLcHE8yyQkqesSqrE9Vc

=== CON IMPLEMENTACIÓN ===

Ferrari: ¡Brrrr!
Trek: Pedalea fuerte
"""

File: JS/Tema9_Ejercicio8.js
/*
 * TEMA 9: PROGRAMACIÓN ORIENTADA A OBJETOS - EJERCICIO 8
 * ========================================================
 * Interfaces (contratos) usando herencia.
 * 
 * EXPLICACIÓN:
 * - Clase abstracta define métodos que deben implementarse
 * - Todas las subclases deben cumplir con la interfaz
 * - Garantizar consistencia
 */

class Pagador {
    pagar() {
        throw new Error("Debe implementar el método pagar()");
    }
}

class TarjetaCredito extends Pagador {
    constructor(numero, cvv) {
        super();
        this.numero = numero;
        this.cvv = cvv;
    }
    
    pagar(cantidad) {
        return `Pagando $${cantidad} con tarjeta ${this.numero.slice(-4)}`;
    }
}

class PayPal extends Pagador {
    constructor(email, contraseña) {
        super();
        this.email = email;
    }
    
    pagar(cantidad) {
        return `Pagando $${cantidad} vía PayPal (${this.email})`;
    }
}

class CriptoMoneda extends Pagador {
    constructor(moneda, billetera) {
        super();
        this.moneda = moneda;
        this.billetera = billetera;
    }
    
    pagar(cantidad) {
        return `Pagando ${cantidad}$ en ${this.moneda} desde billetera: ${this.billetera.slice(0, 10)}...`;
    }
}

class Carrito {
    constructor() {
        this.total = 0;
        this.metodoPago = null;
    }
    
    establecerMetodoPago(pagador) {
        if (pagador instanceof Pagador) {
            this.metodoPago = pagador;
            return true;
        }
        return false;
    }
    
    procesarPago() {
        if (!this.metodoPago) {
            return "Error: no hay método de pago asignado";
        }
        return this.metodoPago.pagar(this.total);
    }
}

console.log("=== INTERFACES ===\n");

const carrito = new Carrito();
carrito.total = 150;

const tarjeta = new TarjetaCredito("1234567890123456", "123");
carrito.establecerMetodoPago(tarjeta);
console.log(carrito.procesarPago());

const paypal = new PayPal("usuario@email.com", "pass");
carrito.establecerMetodoPago(paypal);
console.log(carrito.procesarPago());

const cripto = new CriptoMoneda("Bitcoin", "1A1z7agoat5WFheuQYyvHkQrCHTQiZiM18");
carrito.establecerMetodoPago(cripto);
console.log(carrito.procesarPago());

/*
 * SALIDA ESPERADA:
 * === INTERFACES ===
 * 
 * Pagando $150 con tarjeta 3456
 * Pagando $150 vía PayPal (usuario@email.com)
 * Pagando 150$ en Bitcoin desde billetera: 1A1z7agoa...
 */
